### PR TITLE
[#3567] blueprint feeds controller

### PR DESCRIPTION
**Pull request by tino097**
#Fixes #
#3567

With this PR we are moving feeds controller to feeds Blueprint with all endpoints. Within this change, `webhelpers.FeedGenerator` is replaced by `werkzeug.contrib.atom.AtomFeed` and feed output format is generated by [Atom Syndication](https://tools.ietf.org/html/rfc4287#section-1.1)

### CODE REVIEW

**Comment by amercader** [comment]
This is looking great @tino097! Can't wait to merge all the other Flask stuff to get this properly reviewed and merged

**Comment by amercader** [comment]
@tino097 can you merge latest master please? Cheers

File: ckan/views/feeds.py [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+                      generator=(None, None, None))
+
+    for pkg in results:
+        additional_fields = {}
+
+        for plugin in plugins.PluginImplementations(plugins.IFeed):
+            if hasattr(plugin, u'get_item_additional_fields'):
+                additional_fields = plugin.get_item_additional_fields(pkg)
+
+        feed.add(title=pkg.get(u'title', u''),
```
Comment (amercader) [inline ckan/views/feeds.py]: Please add the following missing parameters:

* `categories`: to include the dataset tags
* `links`: to add a link to the API version of the dataset. See examples in the current feeds, but change the links to the v3 of the API (eg `url_for('api.action', logic_function='package_show', id=...)`)

http://werkzeug.pocoo.org/docs/0.12/contrib/atom/#werkzeug.contrib.atom.FeedEntry

File: ckan/views/feeds.py [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+            feed_class = plugin.get_feed_class()
+
+    if not feed_class:
+        feed_class = _FixedAtomFeed
+
+    feed = feed_class(title=feed_title,
+                      url=feed_link,
+                      language=u'en',
+                      author={u'name': author_name},
+                      author_link=author_link,
```
Comment (amercader) [inline ckan/views/feeds.py]: There is no `author_link`, use `uri` in the `author` dict

File: ckan/views/feeds.py [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+    return _feed_url(search_params, controller=u'package', action=u'search')
+
+
+def _feed_url(query, controller, action, **kwargs):
+    """
+    Constructs the url for the given action.  Encoding the query
+    parameters.
+    """
+    # endpoint = controller + '.' + action
+    path = h.url_for(controller=controller, action=action, **kwargs)
```
Comment (amercader) [inline ckan/views/feeds.py]: Would be good to refactor this to use the new Flask based `url_for` style. Also I'm pretty sure that if we pass the query params as `kwargs` to `url_for` they will get encoded and we don't need to the call below to `_url_with_params`

File: ckan/config/routing.py [hunk 4]
```diff
@@ -372,12 +372,12 @@ def make_map():
         m.connect('/revision/{id}', action='read')
 
     # feeds
-    with SubMapper(map, controller='feed') as m:
-        m.connect('/feeds/group/{id}.atom', action='group')
-        m.connect('/feeds/organization/{id}.atom', action='organization')
-        m.connect('/feeds/tag/{id}.atom', action='tag')
-        m.connect('/feeds/dataset.atom', action='general')
-        m.connect('/feeds/custom.atom', action='custom')
+    # with SubMapper(map, controller='feed') as m:
```
[the remark this answers is not in the document]
Comment (tino097) [reply ckan/config/routing.py]: removed

File: ckan/tests/controllers/test_feed.py [hunk 5]
```diff
@@ -20,6 +18,11 @@ def test_atom_feed_page_zero_gives_error(self):
         offset = url_for(controller='feed', action='group',
```
[the remark this answers is not in the document]
Comment (tino097) [reply ckan/tests/controllers/test_feed.py]: done

File: ckan/tests/controllers/test_feed.py [hunk 6]
```diff
@@ -20,6 +18,11 @@ def test_atom_feed_page_zero_gives_error(self):
...
                          id=group['name']) + '?page=0'
         app = self._get_test_app()
+        with app.flask_app.test_request_context():
```
[the remark this answers is not in the document]
Comment (tino097) [reply ckan/tests/controllers/test_feed.py]: removed

File: ckan/views/feeds.py (hunk 2 above)
Comment (tino097) [reply ckan/views/feeds.py]: done

File: ckan/views/feeds.py (hunk 1 above)
Comment (tino097) [reply ckan/views/feeds.py]: im little bit unsure how to do the part with the links, as the old implementation is using the webhelpers and i didnt find adequate replacement. i could add it with some helper function

File: ckan/views/feeds.py [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+
+    for pkg in results:
+        additional_fields = {}
+
+        for plugin in plugins.PluginImplementations(plugins.IFeed):
+            if hasattr(plugin, u'get_item_additional_fields'):
+                additional_fields = plugin.get_item_additional_fields(pkg)
+
+        feed.add(title=pkg.get(u'title', u''),
+                 url=BASE_URL + h.url_for(str(u'package.read'), id=pkg['id']),
```
[the remark this answers is not in the document]
Comment (tino097) [reply ckan/views/feeds.py]: done

File: ckan/views/feeds.py [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+                 decription=pkg.get(u'notes', u''),
```
[the remark this answers is not in the document]
Comment (tino097) [reply ckan/views/feeds.py]: fixed

File: ckan/views/feeds.py [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+                 updated=h.date_str_to_datetime(pkg.get(u'metadata_modified')),
+                 published=h.date_str_to_datetime(
+                     pkg.get(u'metadata_created')),
+                 unique_id=_create_atom_id(u'/dataset%s' % pkg['id']),
+                 author=pkg.get(u'author', u''),
+                 author_email=pkg.get(u'author_email', u''),
```
[the remark this answers is not in the document]
Comment (tino097) [reply ckan/views/feeds.py]: donde

File: ckan/views/feeds.py (hunk 3 above)
Comment (tino097) [reply ckan/views/feeds.py]: donce

File: ckan/views/feeds.py [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+        """
+        updates = [item['updated'] for item in self.entries
+                   if item['updated'] is not None]
+        if not len(updates):  # delegate to parent for default behaviour
+            return super(_FixedAtomFeed, self).latest_post_date()
+        return max(updates)
+
+
+# Routing
+feeds.add_url_rule(u'/dataset.atom', methods=[u'GET', u'POST'],
```
[the remark this answers is not in the document]
Comment (tino097) [reply ckan/views/feeds.py]: done

**Comment by amercader** [comment]
@tino097 there are unrelated changes in i18n and js files that revert things on master? Do you know what is going on with the PR?

**Comment by tino097** [comment]
@amercader im checking now, cant recall that i had a issue with merging data from master, but i will start new PR

**Comment by tino097** [comment]
corrupted PR, closing

**Pull request closed without merging**